<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Data Preparation Tutorial - Liver Vizgen &mdash; Spatial-live 1.0.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=8d563738"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Data Preparation Tutorial - Kidney Visium" href="kidney_demo.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: lightgray" >

          
          
          <a href="../index.html" class="icon icon-home">
            Spatial-live
              <img src="../_static/spatial-logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation &amp; Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#usage">Usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../guide.html">User Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../guide.html#a-quick-demo">A quick demo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../guide.html#variable-types-and-visual-layers">Variable types and visual layers</a></li>
<li class="toctree-l2"><a class="reference internal" href="../guide.html#input-files">Input files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../guide.html#concise-control-interface">Concise control interface</a></li>
<li class="toctree-l2"><a class="reference internal" href="../guide.html#tooltip-and-image-export">Tooltip and image export</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="kidney_demo.html">Data Preparation Tutorial - Kidney Visium</a><ul>
<li class="toctree-l2"><a class="reference internal" href="kidney_demo.html#Python-virtual-environment-setup">Python virtual environment setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="kidney_demo.html#10X-Visium-spatial-data-preparation">10X Visium spatial data preparation</a></li>
<li class="toctree-l2"><a class="reference internal" href="kidney_demo.html#Preprocessing-and-quality-control">Preprocessing and quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="kidney_demo.html#Highly-variable-genes-and-leiden-clustering">Highly variable genes and leiden clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="kidney_demo.html#Preparing-input-files-for-Spatial-Live">Preparing input files for Spatial-Live</a></li>
<li class="toctree-l2"><a class="reference internal" href="kidney_demo.html#Geometric-Json-file">Geometric Json file</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Data Preparation Tutorial - Liver Vizgen</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Python-virtual-environment-setup">Python virtual environment setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Vizgen-MERFISH-Mouse-Liver-data-preparation">Vizgen MERFISH Mouse Liver data preparation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Base-Image-Preprocessing">Base Image Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Data-preprocessing-and-quality-control">Data preprocessing and quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Spatial-Distributions-of-Cells">Spatial Distributions of Cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Assign-Cell-Types">Assign Cell Types</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Hepatocyte-Zonation:-Peri-Portal-and-Peri-Central-zones">Hepatocyte Zonation: Peri-Portal and Peri-Central zones</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Preparing-input-files-for-Spatial-Live">Preparing input files for Spatial-Live</a></li>
<li class="toctree-l2"><a class="reference internal" href="#A-screen-snapshot-from-Spatial-Live">A screen snapshot from Spatial-Live</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: lightgray" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Spatial-live</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Data Preparation Tutorial - Liver Vizgen</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/liver_demo.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Data-Preparation-Tutorial---Liver-Vizgen">
<h1>Data Preparation Tutorial - Liver Vizgen<a class="headerlink" href="#Data-Preparation-Tutorial---Liver-Vizgen" title="Link to this heading"></a></h1>
<p>This Jupyter notebook serves as a continuous demonstration of the step-by-step process involved in preparing input files for Spatial-Live visualization. In a manner akin to the preceding tutorial centered on kidney data sourced from the 10X Visium platform, this tutorial shifts its focus to liver MERFISH data obtained from the Vizgen platform. This specific dataset measures 347 genes across a substantial &gt;300,000 liver cells in a single mouse liver slice. For the purpose of this demonstration,
we have customized and tailored the notebook, drawing inspiration from the <a class="reference external" href="https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_vizgen_mouse_liver.html">squidpy mouse liver tutorial</a> to align with our objectives.</p>
<section id="Python-virtual-environment-setup">
<h2>Python virtual environment setup<a class="headerlink" href="#Python-virtual-environment-setup" title="Link to this heading"></a></h2>
<p>In order to run this Jupyter Notebook and streamline the process for your convenience, we have compiled several supplementary files including “requirements.txt”, which you can use to easily set up a Python virtual environment.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Please git clone the spatial-live from github,</span>
<span class="c1"># and change the WORKROOTDIR to your own working directory</span>
<span class="o">%</span><span class="k">env</span> WORKROOTDIR=/Users/zhenqingye/projects/spatial-live
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
env: WORKROOTDIR=/Users/zhenqingye/projects/spatial-live
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># You can use the below requirements.txt to create your virtual environment</span>
<span class="o">!</span><span class="w"> </span>tree<span class="w"> </span>-L<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="si">${</span><span class="nv">WORKROOTDIR</span><span class="si">}</span>/quickdemo/
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-blue-intense-fg ansi-bold">/Users/zhenqingye/projects/spatial-live/quickdemo/</span>
├── <span class="ansi-blue-intense-fg ansi-bold">kidney</span>
├── <span class="ansi-blue-intense-fg ansi-bold">liver</span>
└── requirements.txt

2 directories, 1 file
</pre></div></div>
</div>
</section>
<section id="Vizgen-MERFISH-Mouse-Liver-data-preparation">
<h2>Vizgen MERFISH Mouse Liver data preparation<a class="headerlink" href="#Vizgen-MERFISH-Mouse-Liver-data-preparation" title="Link to this heading"></a></h2>
<p>You probably noticed the “liver” folder in the above layout, which contains several files of the Liver1Slice1 dataset from Vizgen’s MERFISH Mouse Liver Map: <a class="reference external" href="https://info.vizgen.com/mouse-liver-access">https://info.vizgen.com/mouse-liver-access</a>. In order to run this tutorial we will download the cell_by_gene.csv and cell_metadata.csv, as well as the image-related files including manifest.json, micron_to_mosaic_pixel_transform.csv, mosaic_DAPI_z0.tif, and mosaic_PolyT_z0.tif. We didn’t include these data files in our git repo. Please follow
the instructions to obtain access to the showcase data and download the data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>tree<span class="w"> </span><span class="si">${</span><span class="nv">WORKROOTDIR</span><span class="si">}</span>/quickdemo/liver/
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-blue-intense-fg ansi-bold">/Users/zhenqingye/projects/spatial-live/quickdemo/liver/</span>
└── <span class="ansi-blue-intense-fg ansi-bold">liver1slice1</span>
    ├── cell_by_gene.csv
    ├── cell_metadata.csv
    └── <span class="ansi-blue-intense-fg ansi-bold">images</span>
        ├── manifest.json
        ├── micron_to_mosaic_pixel_transform.csv
        ├── <span class="ansi-green-intense-fg ansi-bold">mosaic_DAPI_z0.tif</span>
        └── <span class="ansi-green-intense-fg ansi-bold">mosaic_PolyT_z0.tif</span>

2 directories, 6 files
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scipy</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">squidpy</span> <span class="k">as</span> <span class="nn">sq</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">import</span> <span class="nn">rasterio</span>
<span class="kn">from</span> <span class="nn">rasterio.enums</span> <span class="kn">import</span> <span class="n">Resampling</span>
<span class="kn">from</span> <span class="nn">skimage</span> <span class="kn">import</span> <span class="n">transform</span>
<span class="kn">import</span> <span class="nn">cv2</span>
<span class="kn">from</span> <span class="nn">PIL</span> <span class="kn">import</span> <span class="n">Image</span>

<span class="kn">from</span> <span class="nn">copy</span> <span class="kn">import</span> <span class="n">deepcopy</span>

<span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">display</span><span class="o">.</span><span class="n">precision</span><span class="o">=</span><span class="mi">3</span>

<span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scanpy==1.9.5 anndata==0.9.2 umap==0.5.4 numpy==1.23.4 scipy==1.11.2 pandas==2.1.0 scikit-learn==1.3.0 statsmodels==0.14.0 igraph==0.10.8 pynndescent==0.5.10
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># please change the rootdir to your own working directory accordingly</span>
<span class="n">rootdir</span> <span class="o">=</span> <span class="s2">&quot;/Users/zhenqingye/projects/spatial-live/&quot;</span>
<span class="n">library_id</span> <span class="o">=</span> <span class="s2">&quot;liver1slice1&quot;</span>
<span class="n">datadir</span> <span class="o">=</span> <span class="n">rootdir</span> <span class="o">+</span> <span class="s1">&#39;quickdemo/liver/&#39;</span>
</pre></div>
</div>
</div>
</section>
<section id="Base-Image-Preprocessing">
<h2>Base Image Preprocessing<a class="headerlink" href="#Base-Image-Preprocessing" title="Link to this heading"></a></h2>
<p>From Vizgen, a selection of TIF images pertinent to this liver sample has been made available. Specifically, for the purpose of this demonstration, we have chosen to work with the DAPI_z0 and PolyT_z0 images. Given their considerable dimensions, we’ve taken the initiative to downsample and rescale them to more compact sizes using a designated scale factor.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scale_factor</span> <span class="o">=</span> <span class="mi">1</span><span class="o">/</span><span class="mi">16</span>

<span class="n">dapi_tif</span>  <span class="o">=</span> <span class="n">rasterio</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">datadir</span> <span class="o">+</span> <span class="n">library_id</span> <span class="o">+</span> <span class="s2">&quot;/images/mosaic_DAPI_z0.tif&quot;</span><span class="p">)</span>
<span class="n">polyt_tif</span> <span class="o">=</span> <span class="n">rasterio</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">datadir</span> <span class="o">+</span> <span class="n">library_id</span> <span class="o">+</span> <span class="s2">&quot;/images/mosaic_PolyT_z0.tif&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># resample data to target shape</span>
<span class="n">dapi_blue</span> <span class="o">=</span> <span class="n">dapi_tif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">out_shape</span><span class="o">=</span><span class="p">(</span>
            <span class="nb">int</span><span class="p">(</span><span class="n">dapi_tif</span><span class="o">.</span><span class="n">height</span> <span class="o">*</span> <span class="n">scale_factor</span><span class="p">),</span>
            <span class="nb">int</span><span class="p">(</span><span class="n">dapi_tif</span><span class="o">.</span><span class="n">width</span> <span class="o">*</span> <span class="n">scale_factor</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">resampling</span><span class="o">=</span><span class="n">Resampling</span><span class="o">.</span><span class="n">bilinear</span>
<span class="p">)</span>

<span class="n">polyt_green</span> <span class="o">=</span> <span class="n">polyt_tif</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">out_shape</span><span class="o">=</span><span class="p">(</span>
            <span class="nb">int</span><span class="p">(</span><span class="n">polyt_tif</span><span class="o">.</span><span class="n">height</span> <span class="o">*</span> <span class="n">scale_factor</span><span class="p">),</span>
            <span class="nb">int</span><span class="p">(</span><span class="n">polyt_tif</span><span class="o">.</span><span class="n">width</span> <span class="o">*</span> <span class="n">scale_factor</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">resampling</span><span class="o">=</span><span class="n">Resampling</span><span class="o">.</span><span class="n">bilinear</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">dapi_blue</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(</span><span class="n">dapi_blue</span><span class="p">)</span>
<span class="n">polyt_green</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(</span><span class="n">polyt_green</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">normalize</span><span class="p">(</span><span class="n">band</span><span class="p">):</span>
    <span class="n">band_min</span><span class="p">,</span> <span class="n">band_max</span> <span class="o">=</span> <span class="p">(</span><span class="n">band</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">band</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
    <span class="n">range_norm</span> <span class="o">=</span> <span class="p">(</span><span class="n">band</span><span class="o">-</span><span class="n">band_min</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="n">band_max</span> <span class="o">-</span> <span class="n">band_min</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">range_norm</span> <span class="o">*</span> <span class="mi">255</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint8</span><span class="p">)</span>

<span class="n">dapi_blue</span> <span class="o">=</span> <span class="n">normalize</span><span class="p">(</span><span class="n">dapi_blue</span><span class="p">)</span>
<span class="n">polyt_green</span> <span class="o">=</span> <span class="n">normalize</span><span class="p">(</span><span class="n">polyt_green</span><span class="p">)</span>
<span class="n">zero_red</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">dapi_blue</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint8</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">im</span> <span class="o">=</span> <span class="n">cv2</span><span class="o">.</span><span class="n">merge</span><span class="p">((</span><span class="n">zero_red</span><span class="p">,</span> <span class="n">polyt_green</span><span class="p">,</span> <span class="n">dapi_blue</span><span class="o">*</span><span class="mi">2</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># define the alpha and beta</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="mf">1.5</span> <span class="c1"># Contrast control</span>
<span class="n">beta</span> <span class="o">=</span> <span class="mi">50</span> <span class="c1"># Brightness control</span>

<span class="c1"># call convertScaleAbs function</span>
<span class="n">adjusted</span> <span class="o">=</span> <span class="n">cv2</span><span class="o">.</span><span class="n">convertScaleAbs</span><span class="p">(</span><span class="n">im</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span> <span class="n">beta</span><span class="o">=</span><span class="n">beta</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">adjusted</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s1">&#39;lower&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(-0.5, 5988.5, -0.5, 5630.5)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_liver_demo_13_1.png" src="../_images/notebooks_liver_demo_13_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>mkdir<span class="w"> </span>-p<span class="w"> </span><span class="si">${</span><span class="nv">WORKROOTDIR</span><span class="si">}</span>/quickdemo/liver/output
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">img</span> <span class="o">=</span> <span class="n">Image</span><span class="o">.</span><span class="n">fromarray</span><span class="p">(</span><span class="n">adjusted</span><span class="p">)</span>
<span class="n">img</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">datadir</span> <span class="o">+</span> <span class="s1">&#39;output/liver_demo.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>tree<span class="w"> </span><span class="si">${</span><span class="nv">WORKROOTDIR</span><span class="si">}</span>/quickdemo/liver/output/
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-blue-intense-fg ansi-bold">/Users/zhenqingye/projects/spatial-live/quickdemo/liver/output/</span>
└── liver_demo.png

0 directories, 1 file
</pre></div></div>
</div>
</section>
<section id="Data-preprocessing-and-quality-control">
<h2>Data preprocessing and quality control<a class="headerlink" href="#Data-preprocessing-and-quality-control" title="Link to this heading"></a></h2>
<p>Likewise, we will navigate through the conventional single-cell analysis pipeline.</p>
<p>First, we load the cell-by-gene and cell-metadata CSVs and construct an AnnData object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sq</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">vizgen</span><span class="p">(</span><span class="n">datadir</span> <span class="o">+</span> <span class="n">library_id</span><span class="p">,</span>
    <span class="n">counts_file</span><span class="o">=</span><span class="s2">&quot;cell_by_gene.csv&quot;</span><span class="p">,</span>
    <span class="n">meta_file</span><span class="o">=</span><span class="s2">&quot;cell_metadata.csv&quot;</span><span class="p">,</span>
    <span class="n">transformation_file</span><span class="o">=</span><span class="s2">&quot;micron_to_mosaic_pixel_transform.csv&quot;</span><span class="p">,</span>
    <span class="n">library_id</span> <span class="o">=</span> <span class="n">library_id</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 395215 × 347
    obs: &#39;fov&#39;, &#39;volume&#39;, &#39;min_x&#39;, &#39;max_x&#39;, &#39;min_y&#39;, &#39;max_y&#39;
    uns: &#39;spatial&#39;
    obsm: &#39;blank_genes&#39;, &#39;spatial&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;liver1slice1&#39;: {&#39;metadata&#39;: {},
  &#39;scalefactors&#39;: {&#39;transformation_matrix&#39;:        0      1        2
   0  9.259  0.000  385.306
   1  0.000  9.259  999.818
   2  0.000  0.000    1.000}}}
</pre></div></div>
</div>
<p>Configuring pertinent parameters for the spatial imaging entity.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][</span><span class="n">library_id</span><span class="p">][</span><span class="s1">&#39;images&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][</span><span class="n">library_id</span><span class="p">][</span><span class="s2">&quot;images&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;hires&quot;</span><span class="p">:</span> <span class="n">adjusted</span><span class="p">}</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][</span><span class="n">library_id</span><span class="p">][</span><span class="s1">&#39;scalefactors&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s1">&#39;tissue_hires_scalef&#39;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">})</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][</span><span class="n">library_id</span><span class="p">][</span><span class="s1">&#39;scalefactors&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s1">&#39;spot_diameter_fullres&#39;</span><span class="p">:</span> <span class="mi">5</span><span class="p">})</span>
</pre></div>
</div>
</div>
<p>Rescaling and converting micron coordinates to pixel positions based on the adjsted image.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">transmtx</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][</span><span class="n">library_id</span><span class="p">][</span><span class="s1">&#39;scalefactors&#39;</span><span class="p">][</span><span class="s1">&#39;transformation_matrix&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">pixel_positions</span> <span class="o">=</span> <span class="p">[</span> <span class="n">scale_factor</span> <span class="o">*</span> <span class="n">transform</span><span class="o">.</span><span class="n">matrix_transform</span><span class="p">(</span><span class="n">micron</span><span class="p">,</span> <span class="n">transmtx</span><span class="p">)</span> <span class="k">for</span> <span class="n">micron</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span> <span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pixel_positions</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;mt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;mt-&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">qc_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;mt&quot;</span><span class="p">],</span> <span class="n">percent_top</span><span class="o">=</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_counts</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;normalize total&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;log transform&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;scale&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
normalize total
log transform
scale
</pre></div></div>
</div>
<p>Dimensionality Reduction, Neighbor Calculation, and Clustering</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;PCA&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;arpack&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;neighbors&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;UMAP&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PCA
neighbors
UMAP
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">resolution</span> <span class="o">=</span> <span class="mf">1.5</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Leiden&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="n">resolution</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Leiden
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_liver_demo_31_0.png" class="no-scaled-link" src="../_images/notebooks_liver_demo_31_0.png" style="width: 514px; height: 417px;" />
</div>
</div>
</section>
<section id="Spatial-Distributions-of-Cells">
<h2>Spatial Distributions of Cells<a class="headerlink" href="#Spatial-Distributions-of-Cells" title="Link to this heading"></a></h2>
<p>Here we visualize the spatial locations of cells in the mouse liver colored by Leiden cluster.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">(</span> <span class="n">adata</span><span class="p">,</span> <span class="n">img</span><span class="o">=</span><span class="s2">&quot;hires&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;leiden&quot;</span><span class="p">,</span> <span class="n">library_id</span><span class="o">=</span><span class="n">library_id</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">size</span><span class="o">=</span><span class="mi">3</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_liver_demo_33_0.png" class="no-scaled-link" src="../_images/notebooks_liver_demo_33_0.png" style="width: 489px; height: 381px;" />
</div>
</div>
</section>
<section id="Assign-Cell-Types">
<h2>Assign Cell Types<a class="headerlink" href="#Assign-Cell-Types" title="Link to this heading"></a></h2>
<section id="Reference-Cell-Type-Marker-Gene-Sets">
<h3>Reference Cell Type Marker Gene Sets<a class="headerlink" href="#Reference-Cell-Type-Marker-Gene-Sets" title="Link to this heading"></a></h3>
<p>We adhere to the guidelines outlined in the <a class="reference external" href="https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_vizgen_mouse_liver.html">squidpy tutorial</a> to allocate liver cell types, and utilize a gene-level cell type marker reference from the publication titled <a class="reference external" href="https://www.nature.com/articles/s41421-021-00266-1">“Spatial transcriptome profiling by MERFISH reveals fetal liver hematopoietic stem cell niche architecture.”</a> These marker genes serve as pivotal elements in evaluating the
cell type composition of the Leiden clusters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_panel</span> <span class="o">=</span> <span class="s2">&quot;https://static-content.springer.com/esm/art%3A10.1038</span><span class="si">%2F</span><span class="s2">s41421-021-00266-1/MediaObjects/41421_2021_266_MOESM1_ESM.xlsx&quot;</span>
<span class="n">df_ref_panel_ini</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">gene_panel</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">df_ref_panel</span> <span class="o">=</span> <span class="n">df_ref_panel_ini</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">1</span><span class="p">:,</span> <span class="p">:</span><span class="mi">1</span><span class="p">]</span>
<span class="n">df_ref_panel</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">df_ref_panel</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Function&quot;</span><span class="p">]</span>

<span class="c1"># Assign marker gene metadata using reference dataset</span>
<span class="n">marker_genes</span> <span class="o">=</span> <span class="n">df_ref_panel</span><span class="p">[</span>
    <span class="n">df_ref_panel</span><span class="p">[</span><span class="s2">&quot;Function&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;marker&quot;</span><span class="p">)</span>
<span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="n">meta_gene</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
<span class="n">common_marker_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">meta_gene</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">marker_genes</span><span class="p">))</span>
<span class="n">meta_gene</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">common_marker_genes</span><span class="p">,</span> <span class="s2">&quot;Markers&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_ref_panel</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="n">common_marker_genes</span><span class="p">,</span> <span class="s2">&quot;Function&quot;</span>
<span class="p">]</span>
<span class="n">meta_gene</span><span class="p">[</span><span class="s2">&quot;Markers&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta_gene</span><span class="p">[</span><span class="s2">&quot;Markers&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;N.A.&quot;</span> <span class="k">if</span> <span class="s2">&quot;marker&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span>
<span class="p">)</span>
<span class="n">meta_gene</span><span class="p">[</span><span class="s2">&quot;Markers&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Markers
N.A.                            285
HSC marker                        8
Hepatocyte marker                 5
Erythroid progenitor marker       5
Macrophage marker                 5
SEC marker                        5
Neutrophil marker                 5
MK marker                         5
Potential HSC marker              5
Pre-B cell marker                 5
Erythroid cell marker             4
EC marker                         4
AEC marker                        2
Myeloid marker                    2
Macrophage/Hepatocyte marker      1
Basophil marker                   1
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Calculate-Leiden-Cluster-Average-Expression-Signatures">
<h3>Calculate Leiden Cluster Average Expression Signatures<a class="headerlink" href="#Calculate-Leiden-Cluster-Average-Expression-Signatures" title="Link to this heading"></a></h3>
<p>Here we calculate the average gene expression signatures of the Leiden clusters, which will be used to assign cell type composition.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ser_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">ser_counts</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;cell counts&quot;</span>
<span class="n">meta_leiden</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">ser_counts</span><span class="p">)</span>

<span class="n">cat_name</span> <span class="o">=</span> <span class="s2">&quot;leiden&quot;</span>
<span class="n">sig_leiden</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cat_name</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>
<span class="p">)</span>
<span class="k">for</span> <span class="n">clust</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cat_name</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">:</span>
    <span class="n">sig_leiden</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">clust</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cat_name</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="n">clust</span><span class="p">]),</span> <span class="p">:]</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">sig_leiden</span> <span class="o">=</span> <span class="n">sig_leiden</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
<span class="n">leiden_clusters</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Leiden-&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sig_leiden</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
<span class="n">sig_leiden</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">leiden_clusters</span>
<span class="n">meta_leiden</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">sig_leiden</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">meta_leiden</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
    <span class="n">meta_leiden</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">index</span><span class="o">=</span><span class="n">meta_leiden</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Assign-Cell-Type-Based-on-Top-Expressed-Marker-Genes">
<h3>Assign Cell Type Based on Top Expressed Marker Genes<a class="headerlink" href="#Assign-Cell-Type-Based-on-Top-Expressed-Marker-Genes" title="Link to this heading"></a></h3>
<p>Here we assign cell type composition to the Leiden clusters by counting the frequency of cell type marker genes in the top 30 most up-regulated genes for each cluster such that cell type is assigned based on the most frequently occurring marker genes. If there is a tie in the number of marker genes, we assign the cluster to more than one cell type. Using this approach several Leiden clusters are assigned to be Hepatocyte-containing clusters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">meta_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">sig_leiden</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="n">meta_gene</span><span class="p">[</span><span class="s2">&quot;info&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">meta_gene</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="n">meta_gene</span><span class="p">[</span><span class="s2">&quot;Markers&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="s2">&quot;N.A.&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">sig_leiden</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="n">meta_gene</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">common_marker_genes</span><span class="p">,</span> <span class="s2">&quot;Markers&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_ref_panel</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
    <span class="n">common_marker_genes</span><span class="p">,</span> <span class="s2">&quot;Function&quot;</span>
<span class="p">]</span>

<span class="n">meta_leiden</span><span class="p">[</span><span class="s2">&quot;Cell_Type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="s2">&quot;N.A.&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">meta_leiden</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="n">num_top_genes</span> <span class="o">=</span> <span class="mi">30</span>
<span class="k">for</span> <span class="n">inst_cluster</span> <span class="ow">in</span> <span class="n">sig_leiden</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
    <span class="n">top_genes</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">sig_leiden</span><span class="p">[</span><span class="n">inst_cluster</span><span class="p">]</span>
        <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[:</span><span class="n">num_top_genes</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="n">inst_ser</span> <span class="o">=</span> <span class="n">meta_gene</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">top_genes</span><span class="p">,</span> <span class="s2">&quot;Markers&quot;</span><span class="p">]</span>
    <span class="n">inst_ser</span> <span class="o">=</span> <span class="n">inst_ser</span><span class="p">[</span><span class="n">inst_ser</span> <span class="o">!=</span> <span class="s2">&quot;N.A.&quot;</span><span class="p">]</span>
    <span class="n">ser_counts</span> <span class="o">=</span> <span class="n">inst_ser</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

    <span class="n">max_count</span> <span class="o">=</span> <span class="n">ser_counts</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

    <span class="n">max_cat</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">ser_counts</span><span class="p">[</span><span class="n">ser_counts</span> <span class="o">==</span> <span class="n">max_count</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))</span>
    <span class="n">max_cat</span> <span class="o">=</span> <span class="n">max_cat</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; marker&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">inst_cluster</span><span class="p">,</span> <span class="n">max_cat</span><span class="p">)</span>
    <span class="n">meta_leiden</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">inst_cluster</span><span class="p">,</span> <span class="s2">&quot;Cell_Type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">max_cat</span>

<span class="c1"># rename clusters</span>
<span class="n">meta_leiden</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta_leiden</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;Cell_Type&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
<span class="p">)</span>
<span class="n">leiden_names</span> <span class="o">=</span> <span class="n">meta_leiden</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">meta_leiden</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">leiden_names</span>


<span class="c1"># transfer cell type labels to single cells</span>
<span class="n">leiden_to_cell_type</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">meta_leiden</span><span class="p">)</span>
<span class="n">leiden_to_cell_type</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;leiden&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">leiden_to_cell_type</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="kc">None</span>

<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Cell_Type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">leiden_to_cell_type</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;Leiden-&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="s2">&quot;Cell_Type&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Cluster&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">leiden_to_cell_type</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;Leiden-&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="s2">&quot;name&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Leiden-0 SEC
Leiden-1 Hepatocyte
Leiden-2 Hepatocyte
Leiden-3 Hepatocyte
Leiden-4 SEC
Leiden-5 Hepatocyte
Leiden-6 Hepatocyte
Leiden-7 Hepatocyte
Leiden-8 Macrophage
Leiden-9 Hepatocyte
Leiden-10 Macrophage
Leiden-11 AEC_Potential-HSC
Leiden-12 Hepatocyte
Leiden-13 AEC
Leiden-14 MK
Leiden-15 HSC
Leiden-16 Neutrophil
Leiden-17 HSC_Pre-B-cell
Leiden-18 Neutrophil
Leiden-19 HSC_Myeloid
Leiden-20 MK_Pre-B-cell_SEC
Leiden-21 Hepatocyte_Macrophage
Leiden-22 MK_Pre-B-cell_SEC
Leiden-23 SEC
Leiden-24 Macrophage
Leiden-25 SEC
Leiden-26 Erythroid-cell_MK
Leiden-27 Pre-B-cell
Leiden-28 AEC
</pre></div></div>
</div>
</section>
</section>
<section id="Hepatocyte-Zonation:-Peri-Portal-and-Peri-Central-zones">
<h2>Hepatocyte Zonation: Peri-Portal and Peri-Central zones<a class="headerlink" href="#Hepatocyte-Zonation:-Peri-Portal-and-Peri-Central-zones" title="Link to this heading"></a></h2>
<p>Hepatocytes, as the predominant cell type within the liver, fulfill numerous essential functions. They are organized into intricate radial structures known as lobules, each comprising a central vein with lower blood oxygen levels, encompassed by peripheral portal veins characterized by higher blood oxygen levels. This arrangement allows hepatocytes to be broadly categorized into peri-central and peri-portal types based on their proximity to central and portal veins, respectively. Our current
focus lies on these specific zones, leading us to establish a categorical variable delineating their boundaries. Drawing from the <a class="reference external" href="https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_vizgen_mouse_liver.html">squidpy tutorial</a>, we adopt the expression of Axin2 as a marker to distinguish peri-central (Axin2 positive) and peri-portal (Axin2 negative) hepatocytes (refer to <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/31866224/">Sun et. al. 2020</a>). Additionally, for the sake of
simplicity, other non-hepatocyte cells are succinctly grouped under the “peri-other” classification.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_hepatocyte_clusters</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">meta_leiden</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">if</span> <span class="s2">&quot;Hepatocyte&quot;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]</span>
<span class="n">sig_leiden</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">meta_leiden</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">ser_axin2</span> <span class="o">=</span> <span class="n">sig_leiden</span><span class="p">[</span><span class="n">all_hepatocyte_clusters</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;Axin2&quot;</span><span class="p">]</span>
<span class="n">peri_central</span> <span class="o">=</span> <span class="n">ser_axin2</span><span class="p">[</span><span class="n">ser_axin2</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">peri_portal</span> <span class="o">=</span> <span class="n">ser_axin2</span><span class="p">[</span><span class="n">ser_axin2</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="n">peri_zones</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">o</span><span class="p">[</span><span class="s1">&#39;Cluster&#39;</span><span class="p">],</span> <span class="n">peri_central</span><span class="p">):</span>
        <span class="n">peri_zones</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;peri_central&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">o</span><span class="p">[</span><span class="s1">&#39;Cluster&#39;</span><span class="p">],</span> <span class="n">peri_portal</span><span class="p">):</span>
        <span class="n">peri_zones</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;peri_portal&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">peri_zones</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;peri_other&quot;</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;peri_zone&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">peri_zones</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;peri_zone&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
peri_zone
peri_other      156124
peri_portal     137668
peri_central     73443
Name: count, dtype: int64
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">img</span><span class="o">=</span><span class="s2">&quot;hires&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;peri_zone&quot;</span><span class="p">,</span> <span class="n">library_id</span><span class="o">=</span><span class="n">library_id</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">img</span><span class="o">=</span><span class="s2">&quot;hires&quot;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;peri_central&#39;</span><span class="p">,</span> <span class="s1">&#39;peri_portal&#39;</span><span class="p">],</span> <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;peri_zone&quot;</span><span class="p">,</span> <span class="n">library_id</span><span class="o">=</span><span class="n">library_id</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_liver_demo_43_0.png" class="no-scaled-link" src="../_images/notebooks_liver_demo_43_0.png" style="width: 942px; height: 362px;" />
</div>
</div>
<section id="Finding-marker-genes">
<h3>Finding marker genes<a class="headerlink" href="#Finding-marker-genes" title="Link to this heading"></a></h3>
<p>We will proceed to calculate a ranking for highly differential genes within each peri_zone group. Notably, we have omitted the “peri_other” group from this calculation, as it falls outside the scope of our current focus.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;peri_zone&quot;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;peri_central&#39;</span><span class="p">,</span> <span class="s1">&#39;peri_portal&#39;</span><span class="p">],</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;wilcoxon&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_liver_demo_45_0.png" class="no-scaled-link" src="../_images/notebooks_liver_demo_45_0.png" style="width: 817px; height: 440px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">][</span><span class="s1">&#39;names&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peri_central</th>
      <th>peri_portal</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Aldh3a2</td>
      <td>Hsd17b6</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Cyp1a2</td>
      <td>Pck1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Alas1</td>
      <td>G6pc</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Cyp2c38</td>
      <td>Cyp17a1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Axin2</td>
      <td>Hc</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Choose the top one marker from each group for spatial-scatter plottings.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">shape</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Aldh3a2&quot;</span><span class="p">,</span> <span class="s2">&quot;Hsd17b6&quot;</span><span class="p">],</span> <span class="n">library_id</span><span class="o">=</span><span class="n">library_id</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_liver_demo_48_0.png" class="no-scaled-link" src="../_images/notebooks_liver_demo_48_0.png" style="width: 958px; height: 299px;" />
</div>
</div>
<p>Furthermore, we have elected to incorporate the two genes, namely Vwf and Axin2, which were showcased in the squidpy tutorial, as our ultimate input variables for subsequent spatial-live visualization.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">img</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Vwf&quot;</span><span class="p">,</span> <span class="s2">&quot;Axin2&quot;</span><span class="p">],</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Reds&quot;</span><span class="p">,</span> <span class="n">library_id</span><span class="o">=</span><span class="n">library_id</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_liver_demo_50_0.png" class="no-scaled-link" src="../_images/notebooks_liver_demo_50_0.png" style="width: 949px; height: 299px;" />
</div>
</div>
</section>
</section>
<section id="Preparing-input-files-for-Spatial-Live">
<h2>Preparing input files for Spatial-Live<a class="headerlink" href="#Preparing-input-files-for-Spatial-Live" title="Link to this heading"></a></h2>
<p>Now we are ready to extract the necessary data columns for the Spatial-Live input CSV file. This entails a set of essential columns, notably “id:barcode”, “pos:pixel_x”, and “pos:pixel_y”. Additionally, we aim to incorporate one categorical variable, namely “char:peri_zone”, as well as two numerical variables, “num:Vwf” and “num:Axin2”. Furthermore, we’ll designate two additional genes for generating heatmap plots, “gene:Aldh3a2” and “gene:Hsd17b6” from the top markers in previous sections.</p>
<p>Firstly, we extract the relevant columns corresponding to designated genes from the adata object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">csv_out_df</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">obs_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;peri_zone&#39;</span><span class="p">,</span> <span class="s1">&#39;Vwf&#39;</span><span class="p">,</span> <span class="s1">&#39;Axin2&#39;</span><span class="p">,</span> <span class="s1">&#39;Aldh3a2&#39;</span><span class="p">,</span> <span class="s1">&#39;Hsd17b6&#39;</span><span class="p">])</span>
<span class="n">csv_out_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peri_zone</th>
      <th>Vwf</th>
      <th>Axin2</th>
      <th>Aldh3a2</th>
      <th>Hsd17b6</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10482024599960584593741782560798328923</th>
      <td>peri_central</td>
      <td>0.904</td>
      <td>0.422</td>
      <td>1.071</td>
      <td>-1.061</td>
    </tr>
    <tr>
      <th>111551578131181081835796893618918348842</th>
      <td>peri_other</td>
      <td>-0.247</td>
      <td>-0.546</td>
      <td>-0.450</td>
      <td>-0.443</td>
    </tr>
    <tr>
      <th>11173636076188568650971531877509781259</th>
      <td>peri_portal</td>
      <td>-0.247</td>
      <td>-0.546</td>
      <td>0.912</td>
      <td>-1.061</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Then, we extract the positions of each cell (spot) in pixel space coordinating with the employed image. As a result, we will persist in utilizing the down-scaled and adjusted “hires” image to compute the pixel positions of each cell. Additionally, it’s essential to account for disparities in the origin points (left, up) between the image coordinate system and the spatial-live pixel coordinate system (left, bottom).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hires_maxY</span> <span class="o">=</span> <span class="n">adjusted</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="n">csv_out_df</span><span class="p">[</span><span class="s1">&#39;pixel_x&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][:,</span><span class="mi">0</span><span class="p">]</span>
<span class="c1"># translate the origin point (Y-axis) to be consistent with the spatial-lvv pixel coordinate</span>
<span class="n">csv_out_df</span><span class="p">[</span><span class="s1">&#39;pixel_y&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hires_maxY</span> <span class="o">-</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][:,</span><span class="mi">1</span><span class="p">]</span>

<span class="n">csv_out_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>peri_zone</th>
      <th>Vwf</th>
      <th>Axin2</th>
      <th>Aldh3a2</th>
      <th>Hsd17b6</th>
      <th>pixel_x</th>
      <th>pixel_y</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10482024599960584593741782560798328923</th>
      <td>peri_central</td>
      <td>0.904</td>
      <td>0.422</td>
      <td>1.071</td>
      <td>-1.061</td>
      <td>2746.355</td>
      <td>68.591</td>
    </tr>
    <tr>
      <th>111551578131181081835796893618918348842</th>
      <td>peri_other</td>
      <td>-0.247</td>
      <td>-0.546</td>
      <td>-0.450</td>
      <td>-0.443</td>
      <td>2761.803</td>
      <td>68.486</td>
    </tr>
    <tr>
      <th>11173636076188568650971531877509781259</th>
      <td>peri_portal</td>
      <td>-0.247</td>
      <td>-0.546</td>
      <td>0.912</td>
      <td>-1.061</td>
      <td>2709.060</td>
      <td>111.925</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>And we further change the column names to comply with the rules from Spatial-Live, refer to the documents.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">csv_out_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;peri_zone&quot;</span><span class="p">:</span> <span class="s2">&quot;char:peri_zone&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;Vwf&quot;</span><span class="p">:</span> <span class="s2">&quot;num:Vwf&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;Axin2&quot;</span><span class="p">:</span> <span class="s2">&quot;num:Axin2&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;Aldh3a2&quot;</span><span class="p">:</span> <span class="s2">&quot;gene:Aldh3a2&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;Hsd17b6&quot;</span><span class="p">:</span> <span class="s2">&quot;gene:Hsd17b6&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;pixel_x&quot;</span><span class="p">:</span> <span class="s2">&quot;pos:pixel_x&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;pixel_y&quot;</span><span class="p">:</span> <span class="s2">&quot;pos:pixel_y&quot;</span>
                  <span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">csv_out_df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>char:peri_zone</th>
      <th>num:Vwf</th>
      <th>num:Axin2</th>
      <th>gene:Aldh3a2</th>
      <th>gene:Hsd17b6</th>
      <th>pos:pixel_x</th>
      <th>pos:pixel_y</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10482024599960584593741782560798328923</th>
      <td>peri_central</td>
      <td>0.904</td>
      <td>0.422</td>
      <td>1.071</td>
      <td>-1.061</td>
      <td>2746.355</td>
      <td>68.591</td>
    </tr>
    <tr>
      <th>111551578131181081835796893618918348842</th>
      <td>peri_other</td>
      <td>-0.247</td>
      <td>-0.546</td>
      <td>-0.450</td>
      <td>-0.443</td>
      <td>2761.803</td>
      <td>68.486</td>
    </tr>
    <tr>
      <th>11173636076188568650971531877509781259</th>
      <td>peri_portal</td>
      <td>-0.247</td>
      <td>-0.546</td>
      <td>0.912</td>
      <td>-1.061</td>
      <td>2709.060</td>
      <td>111.925</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>To simplify matters, we will narrow down our final dataframe’s scope to concentrate solely on hepatocyte cells, specifically the peri_central and peri_portal groups. However, we also intend to incorporate endothelial cells that exhibit the gene Vwf expression (with Vwf &gt; 0).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># narrow down to hepatocyte cells</span>
<span class="n">filter_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flatnonzero</span><span class="p">(</span><span class="n">csv_out_df</span><span class="p">[</span><span class="s1">&#39;char:peri_zone&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;peri_other&#39;</span><span class="p">)</span>

<span class="c1"># We may also want to include endothelial cells with Vwf &gt; 0</span>
<span class="n">vwf_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flatnonzero</span><span class="p">(</span><span class="n">csv_out_df</span><span class="p">[</span><span class="s1">&#39;num:Vwf&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>

<span class="c1"># merge them together and get the final filtered dataframe</span>
<span class="n">union_idx</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">filter_idx</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">vwf_idx</span><span class="p">))</span>
<span class="n">filter_out_df</span> <span class="o">=</span> <span class="n">csv_out_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">union_idx</span><span class="p">)]</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">filter_out_df</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(224756, 7)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">filter_out_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">datadir</span> <span class="o">+</span> <span class="s2">&quot;output/liver_demo.csv&quot;</span><span class="p">,</span> <span class="n">index_label</span><span class="o">=</span><span class="s2">&quot;id:spot&quot;</span><span class="p">,</span> <span class="n">float_format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%.3f</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Noticed that we have both the liver_demo.csv file and the liver_demo.png image file available for Spatial-Live visualization now.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>tree<span class="w"> </span><span class="si">${</span><span class="nv">WORKROOTDIR</span><span class="si">}</span>/quickdemo/liver/
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-blue-intense-fg ansi-bold">/Users/zhenqingye/projects/spatial-live/quickdemo/liver/</span>
├── <span class="ansi-blue-intense-fg ansi-bold">liver1slice1</span>
│   ├── cell_by_gene.csv
│   ├── cell_metadata.csv
│   └── <span class="ansi-blue-intense-fg ansi-bold">images</span>
│       ├── manifest.json
│       ├── micron_to_mosaic_pixel_transform.csv
│       ├── <span class="ansi-green-intense-fg ansi-bold">mosaic_DAPI_z0.tif</span>
│       └── <span class="ansi-green-intense-fg ansi-bold">mosaic_PolyT_z0.tif</span>
└── <span class="ansi-blue-intense-fg ansi-bold">output</span>
    ├── liver_demo.csv
    └── liver_demo.png

3 directories, 8 files
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>head<span class="w"> </span><span class="si">${</span><span class="nv">WORKROOTDIR</span><span class="si">}</span>/quickdemo/liver/output/liver_demo.csv
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
id:spot,char:peri_zone,num:Vwf,num:Axin2,gene:Aldh3a2,gene:Hsd17b6,pos:pixel_x,pos:pixel_y
10482024599960584593741782560798328923,peri_central,0.904,0.422,1.071,-1.061,2746.355,68.591
11173636076188568650971531877509781259,peri_portal,-0.247,-0.546,0.912,-1.061,2709.060,111.925
116320174062690702928590241926360429752,peri_central,-0.247,2.434,0.962,-1.061,2794.730,57.878
117618209435628285601583396363416663677,peri_portal,-0.247,2.058,1.041,0.015,2758.366,83.566
118681466178454285375254183862628606475,peri_other,9.402,-0.546,-1.476,-1.061,2794.231,102.131
119734422988290309520009455767728512386,peri_central,1.117,-0.546,1.114,-0.587,2704.038,81.650
127234768781846058678605511352431699402,peri_central,-0.247,2.722,1.476,-1.061,2790.465,73.567
131343870818306237824447462929184726576,peri_central,-0.247,-0.546,-0.636,-1.061,2713.909,83.113
134206186392275568276967134217212229277,peri_portal,-0.247,1.536,1.118,-1.061,2757.473,104.000
</pre></div></div>
</div>
</section>
<section id="A-screen-snapshot-from-Spatial-Live">
<h2>A screen snapshot from Spatial-Live<a class="headerlink" href="#A-screen-snapshot-from-Spatial-Live" title="Link to this heading"></a></h2>
<p>Enclosed is a screen snapshot sourced from Spatial-Live, provided as a point of reference. This snapshot aptly demonstrates the tool’s capacity to present a wealth of information through a single figure, featuring multiple layers in a visually compelling manner. In the image below, the light-blue and orange heatmaps correspond to the genes “Aldh3a2” and “Hsd17b6,” respectively. Additionally, the green and purple column bars correspond to the genes “Vmf” and “Axin2,” respectively.</p>
<p>PS: Given the substantial number of elements requiring rendering on the screen, notably each layer housing 228,036 data instances, the combined load of four layers nears 1 million elements. Consequently, it’s possible to encounter slight sluggishness when employing spatial-live, particularly on low-end GPU cards, during the exploration of this liver demo.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">IPython.display</span> <span class="kn">import</span> <span class="n">Image</span> <span class="k">as</span> <span class="n">IPyImage</span>
<span class="n">IPyImage</span><span class="p">(</span><span class="s2">&quot;../images/spatial-live-liver.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_liver_demo_65_0.png" src="../_images/notebooks_liver_demo_65_0.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="kidney_demo.html" class="btn btn-neutral float-left" title="Data Preparation Tutorial - Kidney Visium" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Zhenqing Ye.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>